CIPs/cip-145.md

```diff
@@ -51,17 +59,33 @@ branches are pruned. Since pruning branches that contain Data Events would resul
 a Data Event to contain multiple ancestor events in its `prev` field so that a new merge point Data Event can cover
 multiple events.
 
-1. If a stream is in a diverged state (see events `A` and `B` in fig. 5), we only consider branches that contain valid,
-   uncovered Data Events.
-2. For branches that contain valid, uncovered Data Events, we only consider the first Data Event on a branch after the
-   fork.
-3. The Data Event that is covered by the earliest Time Event wins (see event `A` in fig. 5). 
-4. If two uncovered Data Events are covered by Time Events at the same block height, the Data Event with the lower CID
-   wins.
-
-Using multi-prev Data Events allows us to reduce the number of uncovered events and converge the stream so that there is
-only a single uncovered event, without any data abandoned on pruned branches. The stream's converged/diverged state can
-be determined by looking at the `prev` fields of all the Data Events for that stream.
+1. If a stream is in a diverged state, each uncovered event is a candidate tip.
+2. Branches that do not contain dominant Data Events cannot be the tip.
+3. For branches that contain dominant Data Events, consider the earliest Data Event after a fork point.
+4. The branch with the earliest Data Event becomes the tip.
+5. If multiple earliest Data Events have the same time, then the Data Event with the lowest CID becomes the tip.
+
+For an example of these rules in action, see the figure below:
+![Alt text](../assets/cip-145/rules1.png)
+
+1. Based on rule (1), the stream state has 4 candidate tips, `Time 5`, `Time 6`, `Time 4`, and `Data F`. One of these
+   candidate tips will become the tip of the stream.
+   ![Alt text](../assets/cip-145/rules2.png)
+2. The stream forks between the branches for `Time 1` and `Data A`. Based on rules (3) and (4), the branch for `Data A`
+   is the only branch considered for tip selection.
+   ![Alt text](../assets/cip-145/rules3.png)
+3. This branch later forks into additional branches for `Data E`, `Time 4`, and `Data F`. Based on rules (3) and (4),
+   the branches for `Data E` and `Data F` are the only branches considered for tip selection.
+   ![Alt text](../assets/cip-145/rules4.png)
+4. Based on rule (5), since there is a Time Event for `Data E` but not for `Data F`, only the branch for `Data E` is
```

Comment:
Good catch, yes, it should be rule 4. It is time that made the decision because a Data Event without a Time Event is as if it occurred at time infinity.

> Can we end up in the case where there is a yet-unknown, earlier anchor for the other branch in transit?

Yes, that's possible. If there was an as of yet unknown `Time 5.5` corresponding to `Data F`, then `Time 5.5` becomes the tip.

While a lot less likely in the absence of a malicious CAS attempting a late-publishing attack, it is also possible for example for a `Time 1.5` covering `Data B` to be discovered late. This would rewind the state of the stream, marking `Time 5` the tip.

Having said that, this spec provides a way for the application to resolve such a situation without data loss. A user can decide whether to override the default tip with a new event, while keeping the stream history intact.